File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        return daily(coins, amount);
    }
private:
    int recurse(vector<int>& coins, int idx, int amount, vector<vector<int>>& dp) {
        if(idx >= coins.size() || amount <= 0) {
            if(amount == 0)
                return 0;
            return INT_MAX - 1;
        }
        
        if(dp[amount][idx] != -1)
            return dp[amount][idx];
        
        int notPick = recurse(coins, idx + 1, amount, dp);
        int pick = recurse(coins, idx, amount - coins[idx], dp) + 1;
        
        return dp[amount][idx] = min(pick, notPick);
    }
    int daily(vector<int>& coins, int amount) {
        int count = 0;
        vector<vector<int>> dp(amount + 1, vector<int>(coins.size() + 1, -1));
        int ans = recurse(coins, 0, amount, dp);
        return ans != (INT_MAX - 1) ? ans : -1;
    }
};

File: 322-coin-change/322-coin-change.py
class Solution:
    def anant(self, coins, idx, target, summ, dp):
        
        if idx >= len(coins) or summ >= target:
            if summ == target:
                return 0
            return float('inf')
        
        if dp[idx][summ] != -1:
            return dp[idx][summ]
        
        
        summ += coins[idx] 
        pick = self.anant(coins, idx, target, summ, dp) + 1
        summ -= coins[idx]
        notPick = self.anant(coins, idx + 1, target, summ, dp)
        
        dp[idx][summ] = min(pick, notPick)
        return dp[idx][summ]
        
        
        
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [[-1] * (amount + 1) for _ in range(len(coins) + 1)]
        ans = self.anant(coins, 0, amount, 0, dp)
        
        return ans if ans != float('inf') else -1